django/utils/decorators.py

```diff
@@ -147,33 +176,75 @@ def callback(response):
                     if hasattr(middleware, 'process_response'):
                         return middleware.process_response(request, response)
                 return response
-            return _wrapped_view
-        return _decorator
-    return _make_decorator
 
+            @wraps(view_func)
+            async def _wrapped_view_async(request, *args, **kwargs):
+                if hasattr(middleware, 'process_request'):
+                    result = middleware.process_request(request)
+                    if result is not None:
+                        return result
+                if hasattr(middleware, 'process_view'):
+                    result = middleware.process_view(request, view_func, args, kwargs)
+                    if result is not None:
+                        return result
+                try:
+                    response = await view_func(request, *args, **kwargs)
+                except Exception as e:
+                    if hasattr(middleware, 'process_exception'):
+                        result = middleware.process_exception(request, e)
+                        if result is not None:
+                            return result
+                    raise
+                if hasattr(response, 'render') and callable(response.render):
+                    if hasattr(middleware, 'process_template_response'):
+                        response = middleware.process_template_response(request, response)
+                    # Defer running of process_response until after the template
+                    # has been rendered:
+                    if hasattr(middleware, 'process_response'):
+                        def callback(response):
+                            return middleware.process_response(request, response)
+                        response.add_post_render_callback(callback)
+                else:
+                    if hasattr(middleware, 'process_response'):
+                        return middleware.process_response(request, response)
+                return response
 
-def sync_and_async_middleware(func):
-    """
-    Mark a middleware factory as returning a hybrid middleware supporting both
-    types of request.
-    """
-    func.sync_capable = True
-    func.async_capable = True
-    return func
+            return (
+                _wrapped_view_async
+                if asyncio.iscoroutinefunction(view_func)
+                else _wrapped_view_sync
+            )
+        return _decorator
+    return _make_decorator
 
 
-def sync_only_middleware(func):
+def sync_async_wrapper(
+    view_func, process_request=None, process_response=None, **decorator_kwargs
+):
```

Comment:
Not immediately sure about this function... 🤔

Naming for one... (we can ponder that). 

But more, it's half-encapsulating the pattern we see quite a few times in this PR — but then it's not fully applied. SO... either it's needs abstracting a little so it does apply OR it's not paying the rent, and we should inline it. 
(To think about.)